Fix `repr` on `Period` Types, and `DateTime`, `Date`

fchorney opened the pull request:
Jumping off of this previous PR https://github.com/JuliaLang/julia/pull/30200

It seems the `Period` types face a similar issue. `Year`, `Month`, `Week`, `Day`, `Hour`, `Minute`, `Second`, `Millisecond`, `Microsecond`, and `Nanosecond`
```
julia> using Dates

julia> show(Hour(1))
1 hour
``` 
Realistically this should show the equivalent `repr` type output. `Hour(1)`

I also noticed that the previous PR that is referenced in this one prints out "DateTime" or "Date" regardless of how Dates was imported. `import` vs `using`. 

When using import, you would expect
```
julia> import Dates

julia> show(Dates.Hour(1))
Dates.Hour(1)
```

and with using, you would expect
```
julia> using Dates

julia> show(Hour(1))
Hour(1)
```

fchorney committed:
Fix show output for time/date Period types

fchorney committed:
Clean up date/period output changes

JeffBezanson commented:
Nice, thank you (and welcome)!

Not directly related, but I also notice Dates defines a few methods for `Base.string` and then defines `print(io, x) = print(io, string(x))`, which is redundant. In fact it should only define the `print` methods, then you get `string` for free.

fchorney committed:
Fix all the jldoctests for Dates/Periods

JeffBezanson commented on base/number.jl:
```diff
@@ -314,7 +314,7 @@ julia> oneunit(3.7)
 
 julia> import Dates; oneunit(Dates.Day)
-1 day
+Dates.Day(1)
```
This isn't quite right --- `1 day` should still be printed in the REPL, but `repr` should give `Dates.Day(1)`. The `BitPeriod` type can be removed, and instead the current `show` method should be a method `show(io::IO, ::MIME"text/plain", p::Period)`, and your new method might not be necessary since the fallback 2-argument show will probably print `Dates.Day(1)`.

fchorney replied:
Right, I see what you're saying. I think my latest change should satisfy those requirements, while also satisfying `1 day` being shown in outputs such as in a dataframe.

fchorney committed:
Satisfy Period output formats

- Given `x = Dates.Day(1)`
 - The REPL will print: `1 day`
 - show will print: `Dates.Day(1)`
 - repr will print: `Dates.Day(1)`
 - print will print: `1 day`
 - string will print: `1 day`
 - dataframes will print: `1 day`

omus reviewed (approved):
Nice work 👍

fchorney commented:
Looks like the macOS build failed trying to pull a package
```
Got exception outside of a @test
  failed to clone from https://github.com/JuliaIO/JSON.jl.git, error: GitError(Code:ERROR, Class:Net, curl error: Could not resolve host: github.com)
```

omus commented:
@JeffBezanson does this look good to you?
